controllers/storage.go

```diff
@@ -145,6 +146,84 @@ func (s *Storage) RemoveAllButCurrent(artifact sourcev1.Artifact) ([]string, err
 	return deletedFiles, nil
 }
 
+func (s *Storage) getGarbageFiles(artifact sourcev1.Artifact, n int, ttl time.Duration) []string {
+	localPath := s.LocalPath(artifact)
+	dir := filepath.Dir(localPath)
+	garbageFiles := []string{}
+	filesWithCreatedTs := make(map[time.Time]string)
+	// sortedPaths contain all files sorted according to their created ts.
+	sortedPaths := []string{}
+	now := time.Now().UTC()
+	_ = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		createdAt := info.ModTime().UTC()
+		diff := now.Sub(createdAt)
+		// compare the time difference between now and the time at which the file was created
+		// with the provided ttl. delete if difference is greater than the ttl.
+		expired := diff > ttl
+		if path != localPath && !info.IsDir() && info.Mode()&os.ModeSymlink != os.ModeSymlink && expired {
+			garbageFiles = append(garbageFiles, path)
+		}
+		if !info.IsDir() && info.Mode()&os.ModeSymlink != os.ModeSymlink {
+			filesWithCreatedTs[info.ModTime().UTC()] = path
+		}
+		return nil
+	})
+	creationTimestamps := []time.Time{}
+	for ts := range filesWithCreatedTs {
+		creationTimestamps = append(creationTimestamps, ts)
+	}
+	// sort all timestamps in an ascending order.
+	sort.Slice(creationTimestamps, func(i, j int) bool { return creationTimestamps[i].Before(creationTimestamps[j]) })
+	for _, ts := range creationTimestamps {
+		path, ok := filesWithCreatedTs[ts]
+		if !ok {
+			return nil
+		}
+		sortedPaths = append(sortedPaths, path)
+	}
+
+	for i, path := range sortedPaths {
+		if path != localPath && !stringInSlice(path, garbageFiles) && len(sortedPaths) <= n {
+			// append path to garbageFiles and remove it from sortedPaths
+			garbageFiles = append(garbageFiles, path)
+			sortedPaths = append(sortedPaths[:i], sortedPaths[i+1:]...)
+		}
+	}
+
+	return garbageFiles
```

Comment:
Do we need a loop here?
```suggestion
	if len(sortedPaths) >= n && n >= 0 {
	  return sortedPaths[:len(sortedPaths)-n] 
	}
	
	return sortedPaths[:]
```

Reply:
We need to also return the files that have an expired TTL present in `garbageFiles`. We also need a loop to make sure that we don't accidentally consider the current artifact as garbage.

Reply:
We should be able to achieve both goals with a shorter time complexity (current is O(n^4)?).
At the moment we loop the number of items within that directory 4 times:

1- All files in a given folder
2- All timestamps of the above files
3- All timestamps again for sorting
4- All sorted items

What's thought's on:
a) validating against the localPath on the first loop
b) move `creationTimestamps = append(creationTimestamps, ts)` into the first loop
c) use the slice range to remove loop 4 for both sortedPaths and garbageFiles.

Reply:
Umm, my algorithms knowledge isn't that strong so I might be wrong, but O(n^4) would imply that there are 4 nested loops? Here the _worst_ case scenario is O(n^2) where we check whether `path` is already present in `garbageFiles`. About your suggestions:
a) We do validate against `localPath` in the first loop before appending to `garbageFiles`. The reason why we don't validate before appending to `filesWithCreatedTs` is because we want to make sure that we adhere to `maxItemsToBeRetained` which includes the current artifact.
b) That sounds reasonable.
c) I don't think that's doable since we need to check against `localPath` and if the path already exists in `garbageFiles`, but I'll try and see if it works.

Reply:
For point A, would it be feasible to short circuit that on the first loop/walk? 
As I mentioned on another comment, ideally we should limit the max items processed per garbage collection. We could use the same variable to validate/short-circuit the process.